File: rust_exercises/leetcode/Problem720/Solution.java
public class Solution {
    public String longestWord(String[] words) {
        Map<String, Integer> map = new HashMap<>();
        for (String word: words) {
            if (word.length() == 1) {
                map.put(word, 2);
            } else {
                map.put(word, 0);
            }
        }
        
        String longest = null;
        
        for (String word: map.keySet()) {
            if (isValid(word, map)) {
                if (longest == null || word.length() > longest.length() || (word.length() == longest.length() && word.compareTo(longest) < 0)) {
                    longest = word;
                }
            }
        }
        
        return longest;
    }
    
    private boolean isValid(String word, Map<String, Integer> map) {
        if (!map.containsKey(word)) {
            return false;
        }
        int state = map.get(word);
        switch (state) {
            case 2:
                return true;
            case 1:
                return false;
            default:
                if (isValid(word.substring(0, word.length() - 1), map)) {
                    map.replace(word, 2);
                    return true;
                }
                map.replace(word, 1);
                return false;
        }
    }
}

File: rust_exercises/leetcode/Problem720/solution.rs
use std::collections::HashMap;
use std::cmp::Ordering;

impl Solution {
    pub fn longest_word(words: Vec<String>) -> String {
        let mut map: HashMap<String, i32> = HashMap::new();
        for word in words.iter() {
            if (word.len() == 1) {
                map.insert(word.clone(), 2);
            } else {
                map.insert(word.clone(), 0);
            }
        }
        
        let mut longest = String::new();
        
        for word in words.iter() {
            if is_valid(word, &mut map) {
                if word.len() > longest.len() {
                    longest = word.clone();
                } else if word.len() == longest.len() {
                    match word.cmp(&longest) {
                        Ordering::Less => longest = word.clone(),
                        _ => (),
                    }
                }
            }
        }
        
        longest
    }
}

fn is_valid(word: &str, map: &mut HashMap<String, i32>) -> bool {
    match map.get(word) {
        None | Some(1) => false,
        Some(2) => true,
        Some(_) => if is_valid(&word[..word.len() - 1], map) {
            map.insert(String::from(word), 2);
            true
        } else {
            map.insert(String::from(word), 1);
            false
        },
    }
}
